File: src/fractals/mandelbrot.js
import Immutable from 'immutable';

import { contained } from './constants';


export const name = 'Mandelbrot set';
const defaultEscapeAbs = 2;

function check(c, iterationLimit, escapeAbs = defaultEscapeAbs) {
  let z = c;
  let i = 0;
  while (i < iterationLimit) {
    if (z.abs() >= escapeAbs) {
      return i;
    }

    z = z.mul(z).add(c);
    i += 1;
  }

  return contained;
}

export function makeCheck() {
  return check;
}

export const Parameters = Immutable.Record({
});
export const defaultParameters = new Parameters();

File: src/mandelbrot.rs
use super::complex::Complex;

pub fn check(c: Complex<f64>, iteration_limit: usize, escape_abs: f64) -> usize {
    let mut z = c;
    let mut i = 0;

    while i < iteration_limit {
        if z.abs() >= escape_abs {
            return i;
        }

        z = z.square() + c;
        i += 1;
    }

    iteration_limit
}
